sup3r/models/abstract.py

```diff
@@ -1398,16 +1460,24 @@ def get_single_grad(self,
         loss_details : dict
             Namespace of the breakdown of loss components
         """
-        with tf.device(device_name):
-            with tf.GradientTape(watch_accessed_variables=False) as tape:
-                tape.watch(training_weights)
-
-                hi_res_exo = self.get_high_res_exo_input(hi_res_true)
-                hi_res_gen = self._tf_generate(low_res, hi_res_exo)
-                loss_out = self.calc_loss(hi_res_true, hi_res_gen,
-                                          **calc_loss_kwargs)
-                loss, loss_details = loss_out
-
-                grad = tape.gradient(loss, training_weights)
-
+        with tf.device(device_name), tf.GradientTape(
+                watch_accessed_variables=False) as tape:
+            t0 = time.time()
+            tape.watch(training_weights)
+            self._timing_details['dt:tape.watch'] = time.time() - t0
+            t0 = time.time()
+            hi_res_exo = self.get_high_res_exo_input(hi_res_true)
+            self._timing_details[
+                'dt:get_high_res_exo_input'] = time.time() - t0
+            t0 = time.time()
+            hi_res_gen = self._tf_generate(low_res, hi_res_exo)
+            self._timing_details['dt:_tf_generate'] = time.time() - t0
+            t0 = time.time()
+            loss_out = self.calc_loss(hi_res_true, hi_res_gen,
+                                      **calc_loss_kwargs)
+            self._timing_details['dt:calc_loss'] = time.time() - t0
+            t0 = time.time()
+            loss, loss_details = loss_out
+            grad = tape.gradient(loss, training_weights)
+            self._timing_details['dt:tape.gradient'] = time.time() - t0
```

Comment:
all of these timer variables are kind of gross and not very readable. I tried to look up a better way to do this and I honestly don't see anything great. What about defining a utility fun like this:

```
def timer(fun, *args, **kwargs):
  t0 = time.time()
  out = fun(*args, **kwargs)
  t_elap = time.time() - t0
  return out, t_elap
```

So you could call something like htis:
```
hi_res_exo, t1 = timer(self.get_high_res_exo_input, hi_res_true)
self._timing_details['dt:get_high_res_exo_input'] = t1
```

Not great but do you think there's a better way to do this besides all of the t0 variables? You could even have the function be a self method in a timer class that would log the details to self._timing_details instead of returning t_elap.

Reply:
Yeah I like the class idea